pkg/apis/eventing/v1alpha1/subscription_types.go

```diff
@@ -0,0 +1,204 @@
+	// delete the Subscription and recreate it to point to a different
+	// channel, giving the user more control over what semantics should
+	// be used (drain the channel first, possibly have events dropped,
+	// etc.)
+	From corev1.ObjectReference `json:"from"`
+
+	// Call is reference to (optional) function for processing events.
+	// Events from the From channel will be delivered here and replies
+	// are sent to a channel as specified by the Result.
+	// +optional
+	Call *Callable `json:"call,omitempty"`
+
+	// Result specifies (optionally) how to handle events returned from
+	// the Call target.
+	// +optional
+	Result *ResultStrategy `json:"to,omitempty"`
+}
+
+// Callable specifies the reference to an object that's expected to
+// provide the resolved target of the action. Currently we inspect
+// the objects Status and see if there's a predefined Status field
+// that we will then use to dispatch events to be processed by the target.
+// Currently must resolve to a k8s service or Istio virtual service. Note that
+// in the future we should try to utilize subresources (/resolve ?) to
+// make this cleaner, but CRDs do not support subresources yet, so we need
+// to rely on a specified Status field today. By relying on this behaviour
+// we can utilize a dynamic client instead of having to understand all
+// kinds of different types of objects. As long as they adhere to this
+// particular contract, they can be used as a Target.
+// This ensures that we can support external targets and for ease of use
+// we also allow for an URI to be specified.
+// There of course is also a requirement for the resolved Callable to
+// behave properly at the data plane level.
+type Callable struct {
+	// Only one of these can be specified
+
+	// Reference to an object that will be used to find the target
+	// endpoint.
+	// For example, this could be a reference to a Route resource
+	// or a Knative Service resource.
+	// TODO: Specify the required fields the target object must
+	// have in the status.
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	// +optional
+	Target *corev1.ObjectReference `json:"target,omitempty"`
+
+	// Reference to a 'known' endpoint where no resolving is done.
+	// http://k8s-service for example
+	// http://myexternalhandler.example.com/foo/bar
+	// +optional
+	TargetURI *string `json:"targetURI,omitempty"`
+}
+
+type ResultStrategy struct {
+	// This object must fulfill the Sinkable contract.
+	//
+	// TODO: Specify the required fields the target object must
+	// have in the status.
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	// +optional
+	Target *corev1.ObjectReference `json:"target,omitempty"`
+}
+
+type SubscriptionConditionType string
+
+const (
+	// SubscriptionReady is when the From,Channel and Result have been reconciled successfully.
```

Comment:
done.

Reply:
@evankanderson sure "call" *can* be used as a noun, but IMO "call" is going to be interpreted as a verb by anyone who uses this, especially when combined with "from", which is a preposition and the logical starting point among the components involved thereby setting an imperative context (also as I mentioned earlier, when creating a subscription you are actually subscribing TO a channel, not from).

I think all 3 should be nouns, and not ambiguous ones. I'm personally not even convinced the current "channel" and "subscriber" should change, since those are the most straightforwardly descriptive names for the 2 primary roles. Everything related to the handling of a subscriber's response is a side-effect, except in the (rare) "bridge" case, and for that I think it's best to view as a no-op subscriber since a channel should not be interpreted as the direct subscriber of another channel (even if the "bridge" is logical with no physical resource, it's enforcing the tinker-toy conceptual model). I do see how "replyChannel" or something similar would be a better alternative to "replyTo" since the latter presents the same problem of an imperative context, and if we are going to support dynamic channel resolution in addition to hardwired literal values, then "replyStrategy" could be the top-level as @scothis proposed earlier (with the "mapping" example). 

Sorry if this seems like bike-shedding. I do not view it that way, since these are not names used in an internal implementation. When defining the resources that users interact with, the names of those resources and their properties are among the most important things to get right.